File: Python/1110.py
def binPowMod(x, n, mod):
    if x == 0:
        return 0
    res = 1
    a = x
    while n != 0:
        if (n & 1) == 1:
            res = (res * a) % mod
        a = (a * a) % mod
        n >>= 1
    return res

n, m, y = map(int, input().split())
answer = [x for x in range(m) if binPowMod(x, n, m) == y]
if not answer:
    answer.append(-1)
print(*answer)

File: C++/1110.cpp
#include <algorithm>
#include <iostream>
#include <iterator>
#include <vector>

unsigned binPowMod(unsigned x, unsigned n, unsigned mod) {
  if (x == 0) {
    return x;
  }
  unsigned res = 1;
  unsigned a = x;
  while (n != 0) {
    if ((n & 1) == 1) {
      res = (res * a) % mod;
    }
    a = (a * a) % mod;
    n >>= 1;
  }
  return res;
}

int main() {
  unsigned n;
  unsigned m;
  unsigned y;
  std::cin >> n >> m >> y;

  std::vector<unsigned> numbers;
  for (unsigned x = 0; x < m; ++x) {
    if (binPowMod(x, n, m) == y) {
      numbers.push_back(x);
    }
  }

  if (!numbers.empty()) {
    std::copy(numbers.begin(), numbers.end(),
              std::ostream_iterator<unsigned>(std::cout, " "));
    std::cout << '\n';
  } else {
    std::cout << "-1\n";
  }

  return 0;
}
